File: avancado-intermediario/Java/Q1/Pais.java
import java.util.ArrayList;
import java.util.List;

public class Pais {
    private String codigoIso;
    private String nome;
    private long populacao;
    private double dimensaoKm2;
    private List<Pais> fronteiras;


    
    public Pais(String codigoIso, String nome, double dimensaoKm2) {
        this.codigoIso = codigoIso;
        this.nome = nome;
        this.dimensaoKm2 = dimensaoKm2;
        this.populacao = 0; 
        this.fronteiras = new ArrayList<>();
    }

    public String getCodigoIso() {
        return codigoIso;
    }

    public void setCodigoIso(String codigoIso) {
        this.codigoIso = codigoIso;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public long getPopulacao() {
        return populacao;
    }

    public void setPopulacao(long populacao) {
        this.populacao = populacao;
    }

    public double getDimensaoKm2() {
        return dimensaoKm2;
    }

    public void setDimensaoKm2(double dimensaoKm2) {
        this.dimensaoKm2 = dimensaoKm2;
    }

    public List<Pais> getFronteiras() {
        return fronteiras;
    }

    public boolean ehMesmoPais(Pais outro) {
        if (outro == null) {
            return false;
        }
        if (this.codigoIso == null) {
            return outro.codigoIso == null;
        }
        return this.codigoIso.equalsIgnoreCase(outro.codigoIso);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        Pais outro = (Pais) obj;
        return ehMesmoPais(outro);
    }

    @Override
    public int hashCode() {
        return codigoIso != null ? codigoIso.toLowerCase().hashCode() : 0;
    }

    public void adicionarFronteira(Pais pais) {
        if (pais != null && !this.ehMesmoPais(pais) && !this.ehLimitrofe(pais)) {
            this.fronteiras.add(pais);

            if (!pais.ehLimitrofe(this)) {
                pais.adicionarFronteira(this);
            }
        }
    }

    public boolean ehLimitrofe(Pais pais) {
        if (pais == null) {
            return false;
        }
        for (Pais p : this.fronteiras) {
            if (p.ehMesmoPais(pais)) {
                return true;
            }
        }
        return false;
    }

    public double getDensidadePopulacional() {
        if (this.dimensaoKm2 <= 0) {
            return 0.0;
        }
        return (double) this.populacao / this.dimensaoKm2;
    }


    public List<Pais> vizinhosComuns(Pais outro) {
        List<Pais> comuns = new ArrayList<>();
        if (outro == null) {
            return comuns;
        }
        for (Pais p : this.fronteiras) {
            if (outro.ehLimitrofe(p)) {
                comuns.add(p);
            }
        }
        return comuns;
    }

    
    @Override
    public String toString() {
        return "Pais{" +
                "codigoIso='" + codigoIso + '\'' +
                ", nome='" + nome + '\'' +
                ", populacao=" + populacao +
                ", dimensaoKm2=" + dimensaoKm2 +
                ", fronteiras=" + fronteiras.size() + " vizinho(s)" +
                '}';
    }
}

File: avancado-intermediario/Python/Q1/pais.py
class Pais:
    def __init__(self, codigoIso: str, nome: str, dimensaoKm2: float):
        """código ISO, o nome e a dimensão do país.
        população começa com 0 e a lista de fronteiras começa vazia.
        """
        self.__codigoIso = codigoIso
        self.__nome = nome
        self.__dimensaoKm2 = dimensaoKm2
        self.__populacao = 0
        self.__paisesFronteira = []


    # Getters e Setters
    def getCodigoIso(self) -> str:
        return self.__codigoIso

    def setCodigoIso(self, codigoIso: str):
        self.__codigoIso = codigoIso

    def getNome(self) -> str:
        return self.__nome

    def setNome(self, nome: str):
        self.__nome = nome

    def getPopulacao(self) -> int:
        return self.__populacao

    def setPopulacao(self, populacao: int):
        self.__populacao = populacao

    def getDimensaoKm2(self) -> float:
        return self.__dimensaoKm2

    def setDimensaoKm2(self, dimensaoKm2: float):
        self.__dimensaoKm2 = dimensaoKm2

    # Igualdade semântica

    def equals(self, outroPais: 'Pais') -> bool:
        if not isinstance(outroPais, Pais):
            return False
        return self.__codigoIso == outroPais.__codigoIso

    # Adiciona um país à lista de fronteiras

    def adicionarFronteira(self, pais: 'Pais'):
        for p in self.__paisesFronteira:
            if p.equals(pais):
                return  # Evita duplicatas
        self.__paisesFronteira.append(pais)

    # Verifica se um país é limítrofe

    def ehLimitrofe(self, pais: 'Pais') -> bool:
        for p in self.__paisesFronteira:
            if p.equals(pais):
                return True
        return False

    # Densidade populacional

    def getDensidadePopulacional(self) -> float:
        if self.__dimensaoKm2 == 0:
            return 0.0
        return self.__populacao / self.__dimensaoKm2

    # Vizinhos comuns entre dois países

    def getVizinhosComuns(self, outroPais: 'Pais') -> list:
        vizinhosComuns = []
        for p in self.__paisesFronteira:
            if outroPais.ehLimitrofe(p):
                vizinhosComuns.append(p)
        return vizinhosComuns

    # Representação em string

    def __str__(self) -> str:
        fronteiras = ", ".join(p.getNome() for p in self.__paisesFronteira) if self.__paisesFronteira else 'Nenhuma'
        return (
            f'País: {self.__nome} ({self.__codigoIso})\n'       
            f'  População   : {self.__populacao:,} hab.\n'
            f'  Dimensão    : {self.__dimensaoKm2:,.3f} km²\n'
            f'  Densidade   : {self.getDensidadePopulacional():.2f} hab./km²\n'
            f'  Fronteiras  : {fronteiras}'
        )